### ISSUE TITLE
RFC: Lightweight deletes for MergeTree-family tables.

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#19627]
### Requirements: 
0) we want to delete data from MergeTree-family tables efficiently.
1) The cost of those deletes should be close to the cost of a simple SELECT, which matches the same conditions.
   ```
   SELECT count() WHERE <list of conditions>
   ``` 
   and 
   ```
   DELETE ... WHERE <same list of conditions>
   ```
   should also take a similar amount of time.
2) we can have many `DELETE`s (same limitations on the number of `DELETE`s as for the number of `INSERT`s are ok).
2) `DELETE`s should be synchronous (on a single node), and eventual (go together with replication queue) in a replicated system
3) `DELETE`s should not reprocess significant amounts of data - we don't want to rewrite gigabytes of data to remove a single row. 
   That means we need to store a mask of deleted rows instead of deleting them immediately. Real data deletion should happen eventually during normal background operations (merges etc.).
4) Once we calculate the mask of deleted rows (calculation can be expensive), we should use that mask in a very efficient/close to zero cost during further selects. 
   ```
      SELECT count() FROM table;
      DELETE FROM table WHERE <conditions expensive to calculate>; <-- that may take a lot of time, it's ok
      SELECT count() FROM table; <-- that should be fast (i.e., we don't want to reapply the same conditions with every select)
   ```
5) data parts should be immutable during `DELETE`s - that is 'must-have' for replication.
6) standard SQL syntax

### Concept

Generally, that means that we need 
1) when `DELETE` fired - calculate a mask of deleted rows for every data part (for complicated conditions calculation can be costly)
2) instead of deleting anything, do an insert in some special place the following information
    ```
    data_part_name, mask_of_deleted_rows + may be some meta (like max block number)
    ```
3) selects / merges should check that 'special' place when selecting data, applying it as a row mask (applying the mask should be very cheap)
4) we need to run some upkeeping procedures that  will 
    - consolidate those masks in the background: if frequent `DELETE`s happen, we still want
      to keep a minimum number of masks - ideally one per part, and merge several masks into a single mask efficiently
    - remove the masks for parts that were already merged and deleted - to keep that 'special place' compact. 

![lightweight_deletes1](https://user-images.githubusercontent.com/1549571/105839401-b6867200-5fd1-11eb-9ecf-ac04be77d986.png)

<details>
<summary>Illustration using existing abstactions</summary>

#### Illustration using existing abstactions

**Note**: it's just an illustration. Obviously, the feature should be implemented inside MergeTree engine itself, and live together with that as a single logical table, w/o any extra tables. 

0) we have some `my_table` table with `engine=MergeTree` and we want to run DELETEs there.
1) We need to be able to calculate the offset of the row (from part beginning).
imagine we have it as a virtual column named `_part_offset`
2) we have some special storage to store masks for our `my_table` 
```
CRATE TABLE _deleted_rows_mask.my_table (
   part_name String,
   mask AggregateFunction(groupBitmap, UInt32) 
) engine=AggregatingMergeTree ORDER BY part_name;
```
3) deletes work like 
```
INSERT INTO _deleted_rows_mask.my_table SELECT _part, groupBitmapState( _part_offset) FROM my_table WHERE ... 
```
4) now MergeTree reader should be able to apply those masks while reading the part
```
SELECT * FROM my_table
WHERE (_part, _part_offset) NOT IN
(
   SELECT part_name, arrayJoin( bitmapToArray(  groupBitmapMerge( mask ) ) )
   FROM _deleted_rows_mask.my_table
   GROUP BY part_name
);
```
</details>

#### How mutations work

PARTITION_MINBLOCK_MAXBLOCK_LEVEL + mutation_XXX =  PARTITION_MINBLOCK_MAXBLOCK_LEVEL_XXX

for example:
20100101_1_1_0 + mutation_2.txt = 20100101_1_1_0_2 

Disadvantages:
Even a single row update require 'touching' all the parts (renaming them)
Creates a lot of inactive parts
Makes backups problematic. 
ALTER DELETE is always IO (read/write) and CPU (unpack/pack) heavy

#### Lightweight delete:

get the number in every partition (Replicated) / for a table (nonReplicated) exactly the same as mutation.

Share numbering with mutations: so after mutation nr. 2, the lightweight mutation will get a number nr. 3.  

Works very similar to mutation, and share some invariants, but the result of lightweight mutation is stored 'aside'. 

20100101_1_1_0_2 + lightweight_mutation_3.txt = 20100101_1_1_0_2 + deleted_rows_masks_3.bin ( internally it store a mask for part 20100101_1_1_0_2 )

Obviously, masks need to be stored persistently. (fsyncs?)

So we will have one file deleted_rows_masks_3.bin (mutable, appendable) inside the table data folder for every 'lightweight_mutation'.

Lightweight_mutation mask for lightweight_mutations N+1 includes a mask for lightweight_mutations N (and all previous). 

This way we single lightweight_mutation will write a single file only, and we can keep the last N files. 

The last lightweight_mutation number is always stored in RAM together with other part metainformation.

The last mask **may** also be stored in RAM to make usage of it efficient (questionable - masks for large old parts can be big, so maybe smth like LRU cache + or some smart memory structures may be needed for that).

We also need a nice way to read deleted_rows_masks_3.bin for a particular part, so we may also need smth like deleted_rows_masks_3.idx

So the how that 'lightweight_mutation' work 

1) `DELETE FROM TABLE ... ` fired
2) we get the block number(s) for delete (just like with ALTER DELETE), let's say 100
3) work very similar to mutation (part by part), but instead of writing new parts - it 
   produces `deleted_rows_masks_100.bin`  
   - to achieve speed similar to SELECT - it may be needed to use `max_threads` here. 
4) if we get part 'younger' than 100 (for example from another replica, from a merge scheduled before DELETE) we should
   calculate mask for it when the part is added to a working set.
5) merge on 2 parts ( 20100101_1_1_0 + 20100101_2_2_0 ) which have mutation id 3 and masks in
   deleted_rows_masks_3.bin will produce a part with masks applied, and number in the part name, i.e. 
   `20100101_1_1_0 + 20100101_2_2_0 + deleted_rows_masks_3.bin = 20100101_1_2_1_3 (and no mask).`

Existing invariant for mutations: 
  'every part which has block numbers and mutation_nr younger than last mutation id considered unmutated' get changed to 
  'every part which has block numbers and max( mutation_nr || deleted_rows_masks_last_mutation ) younger than last mutation id considered unmutated'

In replicated:

1) every replica pays the same price for lightweight_mutation and applies it on its own set of parts. 
   state of `deleted_rows_masks_X.bin` can differ on different replicas (and will differ, because set of parts on every replica can differ), but eventually, they will get in sync. (after all known parts with block numbers younger than mutation will be processed). 

2) extend FETCH part to send also latest mask together with part. ( Alternatively: introduce deleted_rows_masks exchange, so if some replica has missing masks for some reason it can request from other replica masks for any of parts existing on that replica.)

### Steps

A. Plain MergeTree implementation

1. Prepare simple in-memory storage for masks + API around (mask getter per part, mask setter per mutation_id)
2. parser for SQL deletes, write lightweight mutation to files (similar to mutationXXX.sql)
3. Create a mutation mode that uses lightweight_mutationXXX.sql and creates masks
4. Make MergeTreeReader aware of deleted rows bitmask on SELECT.
   <-- MVP (deletes w/o persistancy)
5. Add serialization for masks to file, pick the best format supporting appending, checksum validation, and indexing for a mask. (reuse the writers from Log family?), support fsync.
6. Cleanup of obsolete mask files
   <-- + persistancy
7. Review / adjust various places where part checks/manipulation happen (server start, merges, normal mutations, alters, etc).
   <-- cover complex scenarios
8. Consider in-memory optimization for mask storage (LRU cache with fixed size?)
9. Consider multithreaded execution of lightweight mutations
   <-- performance optimizations

B. ReplicatedMergeTree implementation

1. Extend replication queue with a new type of task.
2. Extend part sending to send also mask (preserve backward compatibility for older ClickHouse)
3. Review / adjust various places where part checks/manipulation happen (server start, merges, normal mutations, alters etc).
4. Extend part-exchange to be able to recover masks from other replicas.

<details>
<summary>Implementation details (older, rejected version)</summary>

### Implementation details (older, rejected version)

Declaimer: it's just a concept. Details should be rechecked/reviewed.

Every MergeTree table currently has a set of **data parts** inside (it as a structure defined by a user, and can carry a big amount of data), we want to introduce a new set of **deleted_rows_masks parts** with a fixed and predefined structure and normally it will be tiny. It can be one more 'MergeTreeData' object attached to Storage, or just a special 'service' partition with another structure.  The deletes will be transformed into inserts to **deleted_rows_masks parts** (see below).

We will need to add to MergeTreeReader option to apply that mask automatically (so both selects and merges will work).

In the future, we can also consider adding some optimizations to skip whole ranges during PK analysis. 

#### deleted_rows_mask parts

Those **deleted_rows_mask parts** can live in a subfolder `deleted_rows_masks`, inside table data directory. Or live together with other parts (but with some other partition name). 

Internally it can have a logical structure similar to:
```
(
   part_name String,
   part_state Enum8('active' = 0, 'eol' = 1),  /* once there is scheduled merge for a part we need to store it's state separately */
   delete_ids AggregateFunction(groupArray, UInt32) /* ids of all deletes / may be just max will be enough */
   mask AggregateFunction(groupBitmap, UInt32),
   deleted_timestamp SimpleAggregateFunction(max, DateTime),
) engine=AggregatingMergeTree
ORDER BY (part_name, part_state)
TTL deleted_timestamp + 1 HOUR
```

For storing masks of deleted rows per part roaring bitmaps sounds like a good choice:
 - they are compact if not too many rows were deleted
 - will be able to store up to `MAX_UINT32` offsets for every part (~4 bln, I guess should be enough).
 - do 'bitwise or' automatically (needed to merge multiple deletes into a single row per part)
 - can be slower/bigger when a lot of random rows were deleted, but that sounds acceptable. 

Alternatively, simple arrays can be used.

**deleted_rows_masks parts** should be registered in zookeeper the usual way (just like one more 'service' partition), and share the same/single replication queue (this way, we can preserve the proper sequence of `INSERT`s / `DELETE`s / merges). 

The last state of **deleted_rows_masks** may also live in RAM to avoid reading those parts with every select, so when you add a new part into **deleted_rows_mask** we update that state in RAM.

## DELETE process for ReplicatedMergeTree

1) add DELETE to the replication queue. It stored block_number from every partition (just like mutations)

2) each replica having received a DELETE from the replication queue must apply it to its active set of data parts.

3) Each part "younger" (by its block number) than DELETE must have a mask for this part next to the data part in the version corresponding to the DELETE that occurred. To build that:

    * iterate data parts sequentially 
      
    * for every part with max_block_number lower than DELETE block number we:
        1. first check for non-fetched yet parts **deleted_rows_mask partition** and fetch them. 
        2. if we don't have a delete mask for pair of <part_name, DELETE with number X> in a special place (**deleted_rows_mask partition**) then execute the DELETE to calculate the mask.
        3. if there is a merge for that part in the queue we insert it as 'eol' (merges can use only part_state=active).
    
    * skip parts with min_block_number greater than DELETE block number
     
    * if we have a situation: min_block_number < delete block number < max_block_number, and no mask in **deleted_rows_mask partition** - I guess it should be a logical error / should be just forbidden (by preventing the execution of the merges from the queue out of order with deletes, and by extra logic in merge), see also "merges execution" below. 
    
    * after iterating all the parts form an insert into **deleted_rows_mask partition**. If other replicas already did that - it's ok, deletes are idempotent.
          
    After the DELETE record has been successfully processed by all replicas (and the corresponding masks have been created),
    and all merges which were scheduled before DELETE were finished, then the DELETE record is no longer needed anymore in the replication queue. Now we store the history of deleted rows only as a mask.
    
### execution of merges
* Merge could have been assigned before DELETE or after.
![lightweight_deletes2](https://user-images.githubusercontent.com/1549571/105840084-baff5a80-5fd2-11eb-9e41-e1b37973a101.png)
![lightweight_deletes3](https://user-images.githubusercontent.com/1549571/105840091-bdfa4b00-5fd2-11eb-904b-a8de8ca7b172.png)
    
* Merges start depending not only on the source data parts but also on deletes that were executed BEFORE the merge. So we can not start the merge if one of the source parts is not yet ready, or if some DELETE assigned before the merge is not finished.
![lightweight_deletes4](https://user-images.githubusercontent.com/1549571/105840694-b7b89e80-5fd3-11eb-9077-90fbf71f9b34.png)

* Any merge should read data by applying "on the fly" masks of all DELETEs assigned earlier (i.e. with part_state='active') physically deleting and ignore the masks of those DELETEs that were assigned later than the merge started. Thus, even the lagging merges will produce parts with the same checksums.

* After the end of the merge, before adding a part to the active set, we should make sure that for the new (merged)
part there are masks for all DELETE operation launched AFTER the merge assignment. If some of them are missing - they need to be downloaded or calculated. 

![lightweight_deletes5](https://user-images.githubusercontent.com/1549571/105840820-e9ca0080-5fd3-11eb-904a-521744da49ac.png)

### background operations on deleted_rows_mask partition

We want masks to be compact and cheap to use.
  
* multiple replicas can add the same row to the partition with masks at the same time (no big deal, deletes are idempotent)

* masks for the same part are divided into two groups 
    - related to DELETE operations fired BEFORE the merge, which should end the lifetime of that part (they will be applied during the merge)
    - and related to DELETE operations fired after the merge is assigned (they will not be applied during the merge but will be applied during the selections).

* several lines containing masks for the same part can be combined into one line (bitwise OR on masks) within these two groups.

Example. If 100 deletes happened before the merge was assigned and 25 more after it was assigned. Then you can combine
all 100 in one general mask, and 25 in a separate mask.
 * after the part has been removed (as a result of a merge or similar operation), we can safely remove it and related masks
 * replica can safely exchange parts with masks just like regular parts.
  whatever

</details>

# UPDATEs? 

Executing a 'proper' update requires access to the previous state of the row, which may be not available (or can be outdated) due to eventual consistency on any concrete replica. 

The quite easy thing is a special kind of UPDATE similar to UPSERT (internally DELETE as above + INSERT, w/o accessing the previous state of the row). 

Alternatives: 
- wait for `n-1` replication state (finishing all merges/fetches etc assigned before the update) - it can be very long delays.
- apply on the fly till `n-1` replication state will be reached. That may end up with quite strange behavior, when different rows are affected by same update depending on replication state. Questionable: i guess users expect smth similar to ACID rdbms

**Comment by zhicwu** [comment]
@filimonov, couple of questions:
1. Will lightweight mutation return effected rows? 
2. Will you consider a cheap way for rollback(to previous state/savepoint)?
3. Besides MergeTree, will memory table engine be supported? Will table function like jdbc etc. be enhanced too?

**Comment by filimonov** [comment]
1 - maybe/possible (while due to eventual replication collecting the EXACT number can be problematic at a moment of time).
2 - no. We don't introduce any kind of ACID guarantees here. There is a separate task which hopefully will introduce SOME transactions/operations atomicity in clickhouse. 
3 - out of the scope. But all other scenarios are simpler comparing to ReplicatedMergeTree (which is the main and most complex scenario here).

**Comment by zhicwu** [comment]
> 1 - maybe/possible (while due to eventual replication collecting the EXACT number can be problematic at a moment of time).

Not sure if you'll introduce consistency level and/or hinted handoff like Cassandra, but can we just return effected rows from the server instead of quorum nodes in the cluster? Perhaps replication results can be later saved into a table like system.mutations. IMO, lightweight update/delete should look like insert, which will return effected rows. The information is really needed in jdbc driver(and other clients), so that downstream applications know whether the mutation is a success or not, and if the result is expected.

**Comment by json-lrj** [comment]
the delete / update will obviously tear down the performance  of parallel inserts,  by the way update is very costly for columnar store. Maybe a realtime version of ReplacingMergeTree is more suitable for OALP, just support insert overwrite

**Comment by filimonov** [comment]
Summary of 1st round of discussion:
- concept of converting 'deletes' to 'inserts' sounds ok. 
- storing masks in a separate partition - may be simpler alternatives are possible
- applying masks 
- proposed schema will complicate replication too much.

Options:
- make something simpler based on current mutations (may be like an increment, which will solve only part of existing issues).
- try to fit it in https://github.com/ClickHouse/ClickHouse/issues/13978 and do later.
- look for ways to simplify the replication.

**Comment by Slach** [comment]
@filimonov my 50 cents questions

* what should happen when we execute `ALTER TABLE  ... FREEZE PARTITION` and after its runs, we run the `DELETE statement`?
* what should happen when we execute `UPDATE` on different nodes with different `SET` clauses but with the same conditions for WHERE clause which affect the same rows in `ReplicatedMergeTree` table?

**Comment by filimonov** [comment]
> what should happen when we execute UPDATE on different nodes with different SET clauses but with the same conditions for WHERE clause which affect the same rows in ReplicatedMergeTree table? 

Real updates are out of the scope for now (see last section in the top post). UPSERT will be executed in the order of appearing events in zookeeper. 

> what should happen when we execute ALTER TABLE ... FREEZE PARTITION and after its runs, we run the DELETE statement?

We would need to freeze those DELETE masks too, so i think FREEZE here should just copy the deleted_rows_masks_X.bin +mutation_X.sql for all DELETES started **before** freeze.

**Comment by filimonov** [comment]
> Make MergeTreeReader aware of deleted rows bitmask on SELECT.

That part may be applied on the level of applying indices.

> So we will have one file deleted_rows_masks_3.bin (mutable, appendable) inside the table data folder for every 'lightweight_mutation'.

Some alternatives for that? How to make concurrent writes to that file (queue?) ? 
May be store files per part?  @azat mentioned: we may try to store masks similar to marks / skipping indexes.

**Comment by azat** [comment]
>@azat mentioned: we may try to store masks similar to marks / skipping indexes

Right, let me summarize this briefly.

My proposal is to make deleted masks per part, and make it something like data skipping index, since this will allow filter out completely deleted marks, and this should be significant for SELECT, since this will reduce amount of data that should be read from disk, and this will make it in a "native" way. After filter out completely deleted marks each mark should apply the mask and filter out what left (if any). Marks info can be cached in memory, and per-marks bitmask can be LRU + on disk.

As for the structure of this files it can be some bi-level structure that will have per-mark information only if it is required (i.e. the mark contains deleted and non-deleted rows).

Initial proposal suggest one mutable/appendable file, but this has some concerns like file locking (so parallel parts processing will be complex) and consistency of this file.

Also note that if someone cares about amount of files those DELETE will create (but there will be just 1-2 files per part, not bigger then INSERT does, yes there are users who creates lots of parts but does the complexity worth optimizing DELETE for this case?), later this can be optimized with WAL or similar (and yes it should have structure that will allow to get information for the specific part/mark, but this kind of structure is required for that mutable/appendable file anyway). So instead of writing those markers on per-part basis it can be done on per-table basis with ability to convert it to per-part, just like in-memory/compact/wide format now.

Plus per-part files will make this feature simpler in terms of complexity and review (and this means that it may appears earlier).

**Comment by filimonov** [comment]
> I afraid that those files will be almost always empty

Because of that applying that same way as prewhere may have sense (i doubt that we will be able to skip whole granulas often). And if we will use it as skipping index - we anyway will need to process it as a regular filter, similarly to prewhere. So it seems like that precheck (similar to skipping index) can be an extra optimization (not nessesary for the first version)

**Comment by azat** [comment]
>Yes. I afraid that those files will be almost always empty, while writing them will require to allocate lot of buffers (see #6943 ). Imagine you removed 1 row from the table with 1000 parts. That will require creating 1998 empty files, and 2 files with some data...

Agree this might be a problem but I think that solving it the same way as for MergeTree INSERT/merges is fine, i.e. introduce in-memory list + WAL instead of writing to each part later.

>i doubt that we will be able to skip whole granulas often

Only this will give significant boost for SELECT (i.e. avoid reading that part of data at all), I don't know does it will be often or not, but if someone care about this, this should be taken into consideration.

> So it seems like that precheck (similar to skipping index) can be an extra optimization - which is not nessesary for the first version, where we can support only mode 'similar to prewhere'

Yes this can be done later, but this should be very simple and native (and w/o this it will do extra reading on SELECT, and this will have performance impact in case lots of full granulas/marks will be marked as deleted completely)

**Comment by filimonov** [comment]
> Only this will give significant boost for SELECT (i.e. avoid reading that part of data at all), I don't know does it will be often or not, but if someone care about this, this should be taken into consideration.

Merge / optimize final will resolve that problem in a natural way. 

I agree that this may be improvement sometimes, but can also introduce some extra costs (we will need to process/analize masks twice - once to check if we can skip whole granula and second time - on 'prewhere' stage). I would leave that optimization for a later phases: test scenarios need to be modeled / benchmarked.

**Comment by ryan-mei** [comment]
@filimonov @myrrc Nice job, this feature is strong expectations!  Is there any plan about when this feature will be available or testable?

**Comment by spongedu** [comment]
> @filimonov @myrrc Nice job, this feature is strong expectations! Is there any plan about when this feature will be available or testable?

Ditto. Is There any estimated time for this excellent feature ?

**Comment by heyciao** [comment]
> BTW: it looks like we may also need to take a number of deleted rows in the part into account when scheduling the merge. Or schedule some special type of merges to remove deleted rows if we collected a lot of them (like 25% of rows in the part).

It would be very nice to be able to define such a threshold ("schedule merges for parts that have at least X% of deleted rows"). Maybe that could be set globally (valid for all tables), and as well per table (e.g. small tables that are meant to host temporary data or data that can change frequently through "upsert" might need a higher threshold than other normal static ones, to avoid that their parts keep getting merged too frequently?).

Maybe a special merge mode would have to be handled as well, during which the merge rebuilds only the single parts which exceed the threshold? (meaning that a single part would just be rewritten, but without its deleted rows)
I'm thinking about this in relation to "max_bytes_to_merge_at_max_space_in_pool": if a part has reached that limit (e.g. 150GB) it should still be possible to shrink it if it has rows which were deleted... .

**Comment by xjewer** [comment]
Lightweight deletes: https://github.com/ClickHouse/ClickHouse/pull/24755

**Comment by weeds085490** [comment]
# light weight delete design
## 1. Overview of the lightweight deletion process
<img width="988" alt="image" src="https://user-images.githubusercontent.com/13234323/158527074-e59d6b2f-64b4-42c3-b95d-03918cb17f1c.png">


## 2. Implementation details

**DELETE(Reuse the native mutation mechanism)**:
- For each Part, calculate the `row offset`(The row offset of the first row of part data is 0, the second row is 1, and so on) of the hit rows based on the predicate condition of the delete operation.
- For _delete column, mark the corresponding data  as 1 (1 means delete, 0 means valid) according to the row offset. For other columns, create a hard link.

**SELECT**：
- Add _delete != 1 to the predicate condition of the query.

**MERGE**:
- Filter out data rows with _delete = 1.

## 3. Some other optimizations
- Use lightweight deletion to optimize the data rewriting mechanism of MaterializedMySQL, so that we are no longer troubled by FINAL when querying.
- Mutation monitoring and rejection mechanism.
- Increase the priority of lightweight delete operations on Merge-mutate task selection.
- ...

**Comment by alexey-milovidov** [comment]
@weeds085490 Thank you! This design looks exactly as I imagined it. It is reasonable and easy to understand.

**Comment by alexey-milovidov** [comment]
And the `_delete` column can be not written at all if it is all zeros (on new part or after a normal merge).

**Comment by den-crane** [comment]
may be expose _delete as a virtual column for introspection and remove `_delete <>1` if this column is queried ?

**Comment by weeds085490** [comment]
> And the `_delete` column can be not written at all if it is all zeros (on new part or after a normal merge).

Thanks, we have implemented this point. This feature has been working in our production environment for a while. This is our first PR: https://github.com/ClickHouse/ClickHouse/pull/35318. PTAL

**Comment by weeds085490** [comment]
> may be expose _delete as a virtual column for



> may be expose _delete as a virtual column for introspection and remove `_delete <>1` if this column is queried ( or allow to query `where _delete =1` )

Thank you!  That's how we do it

**Comment by WillMeng** [comment]
That's great! Thanks very much.
I have another question, how to do about partial column update？

**Comment by filimonov** [comment]
There are some downsides of that approach, the main one is that every delete in that scenario = changing all the parts during a single DELETE.

If we will name it a usual DELETE - it will not sound like something you should be careful with. DELETEs for all RDBMs are light operations, and you can have a lot of them per sec.

With that approach 
   - a lot of outdated parts will be created (so for example if you will do 1 delete per second - you will have about 480 (hardlink) copies of all parts ( 8 (min old parts lifetime) * 60 (sec) *1 (delete per sec) = 480 ) - high FS load (especially if you have lot of parts / and columns).
   - a lot of zookeeper operations - high Zookeeper load

That concept above was mainly about trying to make the parts themself immutable but store some extra state aside. 

But of course that proposal is simpler & still bring the improvement. So I'm totally ok with that. But may be it would be better to still call it alter. (mayby `ALTER table LIGHTWEIGHT DELETE ...` like in the PR).

**Comment by MSchutt** [comment]
What is the general timeline for implementing lightweight delete & update features? :)

**Comment by alexey-milovidov** [comment]
@MSchutt Q4 2022.
But please ignore everything described in this task, it is no longer relevant.

**Comment by alexon1234** [comment]
> @MSchutt Q4 2022. But please ignore everything described in this task, it is no longer relevant.

Is there any place where we can find more information about it? Thanks

**Comment by alexey-milovidov** [comment]
@alexon1234 a column with delete byte mask that is used to filter data on reading similar to PREWHERE.

@davenger is doing it.

**Comment by alexey-milovidov** [comment]
Merged: #37893
